rplugin/python3/molten/jupyter_server_api.py

```diff
@@ -0,0 +1,134 @@
+
+
+class JupyterAPIClient:
+    def __init__(self,
+                 url: str,
+                 kernel_info: Dict[str, Any],
+                 headers: Dict[str, str]):
+        self._base_url = url
+        self._kernel_info = kernel_info
+        self._headers = headers
+
+        self._recv_queue: Queue[Dict[str, Any]] = Queue()
+
+    def get_stdin_msg(self, **kwargs):
+        return None
+
+    def wait_for_ready(self, timeout: float = 0.):
+        start = time.time()
+        while True:
+            response = requests.get(self._kernel_api_base,
+                                    headers=self._headers)
+            response = json.loads(response.text)
+
+            if response["execution_state"] != "idle" and time.time() - start > timeout:
+                raise RuntimeError
+
+            # Discard unnecessary messages.
+            while True:
+                try:
+                    response = self.get_iopub_msg()
+                except EmptyQueueException:
+                    return
+
+
+    def start_channels(self) -> None:
+        parsed_url = urlparse(self._base_url)
+        self._socket = websocket.create_connection(f"ws://{parsed_url.hostname}:{parsed_url.port}"
+                                                   f"/api/kernels/{self._kernel_info['id']}/channels",
+                                                   header=self._headers,
+                                                   )
+        self._kernel_api_base = f"{self._base_url}/api/kernels/{self._kernel_info['id']}"
+
+        self._iopub_recv_thread = Thread(target=self._recv_message)
+        self._iopub_recv_thread.start()
+
+    def _recv_message(self) -> None:
+        while True:
+            response = json.loads(self._socket.recv())
+            self._recv_queue.put(response)
+
+    def get_iopub_msg(self, **kwargs):
+        if self._recv_queue.empty():
+            raise EmptyQueueException
+
+        response = self._recv_queue.get()
+
+        return response
+
+    def execute(self, code: str):
+        header = {
+            'msg_type': 'execute_request',
+            'msg_id': uuid.uuid1().hex,
+            'session': uuid.uuid1().hex
+        }
+
+        message = json.dumps({
+            'header': header,
+            'parent_header': header,
+            'metadata': {},
+            'content': {
+                'code': code,
+                'silent': False
+            }
+        })
+        self._socket.send(message)
+
+    def shutdown(self):
+        requests.delete(self._kernel_api_base,
+                        headers=self._headers)
+        self._socket.close()
+
+    def cleanup_connection_file(self):
+        pass
+
+class JupyterAPIManager:
+    def __init__(self,
+                 url: str,
+                 ):
+        parsed_url = urlparse(url)
+        self._base_url = f"{parsed_url.scheme}://{parsed_url.netloc}"
+
+        token_part = re.search(r"token=(.*)", parsed_url.query)
```

Comment:
I know it's not common, but wouldn't this fail for a url that had multiple query params?

Can we get this information from `parsed_url`? Or just make this query search up to `$` or `&`